codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -254,6 +260,72 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
         }
     }
 
+    /**
+     * {@inheritDoc}
+     *
+     * Override of {@link MessageToByteEncoder#write(ChannelHandlerContext, Object, ChannelPromise)} that accounts for
```

Comment:
I think we can reduce the code duplication and avoid API breakage by introducing the following overloaded `encode()` method on `MessageToByteEncoder`

```java

    protected ChannelPromise encode(ChannelHandlerContext ctx, I msg, ByteBuf out, ChannelPromise promise)
            throws Exception {
        encode(ctx, msg, out);
        return promise;
    }
```

And instead of calling the previous `encode()` method call this one in `MessageToByteEncoder#write()` like:

`promise = encode(ctx, cast, buf, promise);`

Individual encoders when required can then override this new `encode()` method to modify the `promise`. Similar to what you are doing in this change for `Lz4FrameEncoder`:

```java

    @Override
    protected ChannelPromise encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out, ChannelPromise promise)
            throws Exception {
        try {
            encode(ctx, msg, out);
        } catch (Exception e) {
            if (bufferPromise != null) {
                bufferPromise.tryFailure(e);
            }
            throw e;
        }

        if (bufferPromise == null) {
            if (finished) {
                return promise;
            }
            bufferPromise = promise;
            return ctx.newPromise();
        } else {
            bufferPromise.addListener(new ChannelPromiseNotifier(promise));
            return finished ? bufferPromise : ctx.newPromise();
        }
    }
```

Reply:
Thanks @NiteshKant this looks much nicer indeed. I had to do the new LZ4 encode a little differently though to address the various combinations of not retaining any buffered bytes if the message fits exactly and to switch out the `bufferPromise` if we flush out some bytes but also retain some in `buffer`.

Let me know what you think :)